convert2rhel/unit_tests/actions/actions_test.py

```diff
@@ -0,0 +1,793 @@
+
+    #
+    # Test that Stage.run() works as expected
+    #
+    @pytest.mark.parametrize(
+        ("stage_dirs", "expected"),
+        (
+            # Single Stages
+            (
+                ("good_deps1",),
+                (("REALTEST", "SECONDTEST", "THIRDTEST", "FOURTHTEST"), (), ()),
+            ),
+            # Test that all action statuses are categorized correctly
+            (
+                ("all_status_actions",),
+                (
+                    ("SUCCESSTEST", "WARNINGTEST"),
+                    ("ERRORTEST", "OVERRIDABLETEST"),
+                    ("SKIPSINGLETEST", "SKIPMULTIPLETEST"),
+                ),
+            ),
+            # Test that exceptions inside of Actions are detected as errors
+            (
+                ("action_exceptions",),
+                (("SUCCESSTEST",), ("DIVIDEBYZEROTEST", "LOGCRITICALTEST"), ()),
+            ),
+            # Multiple Stages
+            (
+                ("deps_on_1", "good_deps1"),
+                (("REALTEST", "SECONDTEST", "THIRDTEST", "FOURTHTEST", "TESTI", "TESTII"), (), ()),
+            ),
+        ),
+    )
+    def test_run(self, stage_actions, stage_dirs, expected):
+        stage = None
+        for stage_dir in stage_dirs:
+            stage = actions.Stage(stage_dir, next_stage=stage)
+
+        actual = stage.run()
+
+        # Assert that the success, fail, and skipped lists are the same
+        assert sorted(action.id for action in actual.successes) == sorted(expected[0])
+        assert sorted(action.id for action in actual.failures) == sorted(expected[1])
+        assert sorted(action.id for action in actual.skips) == sorted(expected[2])
+
+    def test_stages_cannot_be_run_twice(self, stage_actions):
+        """Test that an Action can only be run once."""
+        stage = actions.Stage("good_deps1")
+        stage.run()
+
+        with pytest.raises(actions.ActionError, match="Stage good_deps1 has already run"):
+            stage.run()
+
+
+class TestResolveActionOrder:
+    @pytest.mark.parametrize(
+        ("potential_actions", "ordered_result"),
+        (
+            ([], []),
+            ([_ActionForTesting(id="One")], ["One"]),
+            ([_ActionForTesting(id="One"), _ActionForTesting(id="Two", dependencies=("One",))], ["One", "Two"]),
+            ([_ActionForTesting(id="Two"), _ActionForTesting(id="One", dependencies=("Two",))], ["Two", "One"]),
+            (
+                [
+                    _ActionForTesting(id="One"),
+                    _ActionForTesting(id="Two", dependencies=("One",)),
+                    _ActionForTesting(id="Three", dependencies=("Two",)),
+                    _ActionForTesting(id="Four", dependencies=("Three",)),
+                ],
+                ["One", "Two", "Three", "Four"],
+            ),
+            # Multiple dependencies (Slight differences in order to show
+            # that order of deps and Actions doesn't matter.  The sort is
+            # still stable.
```

Comment:
Not a very useful suggestion... but still, it had to be closed. 
```suggestion
            # Multiple dependencies (Slight differences in order to show
            # that order of deps and Actions doesn't matter).  The sort is
            # still stable.
```